storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3088,6 +3119,690 @@ std::ostream &fseg_header::to_stream(std::ostream &out) const
+    {
+      offsets= rec_get_offsets(rec, m_index, offsets,
+                               page_is_leaf(page)
+                               ? m_index->n_core_fields : 0,
+                               ULINT_UNDEFINED, &heap);
+      child_page_no= btr_node_ptr_get_child_page_no(rec, offsets);
+    }
+    else
+    {
+      child_page_no= FIL_NULL;
+      err= DB_CORRUPTION;
+      break;
+    }
+    if (cur_level == level + 1)
+      break;
+  }
+  mtr.commit();
+
+  if (heap)
+    mem_heap_free(heap);
+  return child_page_no;
+}
+
+dberr_t indexDefragmenter::defragment_level(uint16_t level,
+                                            spaceDefragmenter *space_defrag)
+{
+  dberr_t err= DB_SUCCESS;
+  uint32_t cur_page_no= get_level_block(level, &err);
+  if (cur_page_no == FIL_NULL)
+    return err;
+
+  fil_space_t *space= fil_system.sys_space;
+  mtr_t mtr;
+  mtr.start();
+  mtr.x_lock_space(space);
+  mtr_x_lock_index(m_index, &mtr);
+
+  buf_block_t *block= buf_page_get_gen(page_id_t{0, cur_page_no},
+                                       0, RW_X_LATCH, nullptr,
+                                       BUF_GET_POSSIBLY_FREED, &mtr, &err);
+  if (!block)
+  {
+err_exit:
+    mtr.discard_modifications();
+    mtr.commit();
+    return err;
+  }
+
+  while (block)
+  {
+    page_t *page= buf_block_get_frame(block);
+    uint32_t next_page_no= btr_page_get_next(page);
+    uint32_t cur_extent= (cur_page_no / FSP_EXTENT_SIZE) * FSP_EXTENT_SIZE;
+    uint32_t state= space_defrag->get_state(cur_extent);
+
+    if (state == XDES_FREE)
+    {
+fetch_next_page:
+      mtr.commit();
+
+      if (next_page_no == FIL_NULL)
+        break;
+      cur_page_no= next_page_no;
+
+      mtr.start();
+      mtr_x_lock_index(m_index, &mtr);
+      mtr.x_lock_space(fil_system.sys_space);
+      block= buf_page_get_gen(page_id_t{0, cur_page_no}, 0, RW_X_LATCH,
+                            nullptr, BUF_GET_POSSIBLY_FREED, &mtr,
+                            &err);
+      continue;
+    }
+
+    uint32_t new_extent= space_defrag->get_new_extent(cur_extent);
+
+    if (new_extent == cur_extent)
+    {
+next_page:
+      /* Read all child pages exist in parent block */
+      if (level)
+        store_child_pages(block);
+      goto fetch_next_page;
+    }
+
+    /* Need to replace the current extent with new extent */
+    buf_block_t *root= btr_root_block_get(m_index, RW_SX_LATCH, &mtr,
+                                          &err);
+    if (UNIV_UNLIKELY(!root))
+      goto err_exit;
+
+    buf_block_t *new_block= nullptr;
+    fseg_header_t *seg_header=
+      root->page.frame + (level
+                          ? PAGE_HEADER + PAGE_BTR_SEG_TOP
+                          : PAGE_HEADER + PAGE_BTR_SEG_LEAF);
+
+    buf_block_t *iblock;
+    fseg_inode_t *inode= fseg_inode_try_get(seg_header, 0, 0, &mtr,
+                                            &iblock, &err);
+    if (!inode)
+      goto err_exit;
+
+    switch(state)
+    {
```

Comment:
There should be no line break before the `switch` expression and the `{` that starts the body.